### PR TITLE
Reveal some mission target locations on start

### PR DESCRIPTION
**Pull request by patternoia**
Related to #15979 . Though Refugee Center is now revealed, this PR will add the revelation of target locations to these four missions:
- Find Weather Log
- Find Corporate Accounts
- Find Patient Records
- Find Relic

as well as API to search for random overmap type in `overmapbuffer` and some helper functions to `mission_start`

### CODE REVIEW

**Comment by BorkBorkGoesTheCode** [comment]
The computer terminal actually displays a path to the Refugee Center, so I don't think this is necessary. Could a similar system be added to questgivers?

**Comment by DangerNoodle** [comment]
It would be more useful if this feature was converted to JSON like some other quest properties, but it might be harder to implement.

**Comment by patternoia** [comment]
@BorkBorkGoesTheCode this is exactly the case.
The reveals are triggered after getting a quest from NPC

**Comment by patternoia** [comment]
@DangerNoodle I thought about the same, but did not see an elegant way to implement it right now.

**Comment by DangerNoodle** [comment]
> @DangerNoodle I thought about the same, but did not see an elegant way to implement it right now.

I had considered the same, but do not know for sure how the effectively implement it. I had also wanted to look into how to convert the "kill single monster or multiple monsters" quest, but that could also be awkward to implement.

**Comment by BorkBorkGoesTheCode** [comment]
>@BorkBorkGoesTheCode this is exactly the case.
 The reveals are triggered after getting a quest from NPC

Oh. I didn't know that.

File: src/mission_start.cpp [hunk 1]
```diff
@@ -1627,3 +1627,92 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+
+    const tripoint destination = reveal_destination( "cathedral_b", true );
+    if( destination != overmap::invalid_tripoint )
+    {
+        add_msg( _( "%s have marked the only religious building known to them on your map" ), p->name.c_str() );
+    if ( one_in( 3 ) )
+        reveal_route( miss, destination );
+    } else {
+        add_msg( _( "You can't even guess where this religious building could be..." ) );
+    }
```
Comment (codemime) [inline src/mission_start.cpp]: These functions are almost similar. Try to eliminate unnecessary duplication.

File: src/mission_start.cpp [hunk 2]
```diff
@@ -1627,3 +1627,92 @@ void mission_start::ranch_bartender_4(mission *miss)
 void mission_start::place_book( mission *)
 {
 }
+
+const tripoint reveal_destination ( const std::string& type, bool random )
```
Comment (codemime) [inline src/mission_start.cpp]: `reveal_destination( ..., false )` is never invoked.

File: src/mission_start.cpp [hunk 3]
```diff
@@ -1627,3 +1627,92 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+}
+
+void reveal_route( mission *miss, const tripoint destination )
+{
+    const tripoint source = g->u.global_omt_location();
+
+    const tripoint source_road = overmap_buffer.find_closest( source, "road", 3, false );
+    const tripoint dest_road = overmap_buffer.find_closest( destination, "road", 3, false );
+
+    if( overmap_buffer.reveal_route( source_road, dest_road ) )
```
Comment (codemime) [inline src/mission_start.cpp]: Missing/misplaced curly braces (here and below). See [CODE_STYLE.txt](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CODE_STYLE.txt) for details.

File: src/mission_start.cpp [hunk 4]
```diff
@@ -1627,3 +1627,92 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+}
+
+void mission_start::reveal_weather_station( mission *miss )
+{
+    npc *p = g->find_npc( miss->get_npc_id() );
+
+    const tripoint destination = reveal_destination( "station_radio", true );
+    if( destination != overmap::invalid_tripoint )
+    {
+        add_msg( _( "%s have marked the only broadcasting station known to them on your map" ), p->name.c_str() );
```
Comment (codemime) [inline src/mission_start.cpp]: Here and below: `find_npc()` can return `nullptr`. I'm aware that the code was copied from the existing, but the possibility exists.

File: src/mission_start.cpp [hunk 5]
```diff
@@ -1627,3 +1627,92 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+    const tripoint source_road = overmap_buffer.find_closest( source, "road", 3, false );
+    const tripoint dest_road = overmap_buffer.find_closest( destination, "road", 3, false );
+
+    if( overmap_buffer.reveal_route( source_road, dest_road ) )
+        add_msg( _( "%s marks as well the road that leads to it..." ), g->find_npc( miss->get_npc_id() )->name.c_str() );
+}
+
+void mission_start::reveal_weather_station( mission *miss )
+{
+    npc *p = g->find_npc( miss->get_npc_id() );
```
Comment (codemime) [inline src/mission_start.cpp]: Can be `const`.

File: src/overmapbuffer.cpp [hunk 6]
```diff
@@ -582,6 +582,13 @@ std::vector<tripoint> overmapbuffer::find_all( const tripoint& origin, const std
     return result;
 }
 
+tripoint overmapbuffer::find_random( const tripoint& origin, const std::string& type,
+                                     int dist, bool must_be_seen )
+{
+    std::vector<tripoint> result = find_all( origin, type, dist, must_be_seen );
+    return random_entry ( result, overmap::invalid_tripoint );
```
Comment (codemime) [inline src/overmapbuffer.cpp]: Looks better when inlined: `random_entry( find_all( ... ), ...`. This way it doesn't depend on the type returned by `find_all()`. `find_random( ...., false )` is never invoked as well.

**Comment by Coolthulhu** [comment]
Works.

Just needs the above cleanup applied and it's good to go.

**Comment by Coolthulhu** [comment]
This is almost ready, only needs code cleanup.

**Comment by patternoia** [comment]
Sorry for a delay.
I've updated the code

File: src/mission_start.cpp [hunk 7]
```diff
@@ -1627,3 +1627,82 @@ void mission_start::ranch_bartender_4(mission *miss)
 void mission_start::place_book( mission *)
 {
 }
+
+const tripoint reveal_destination( const std::string &type, bool random )
```
Comment (codemime) [inline src/mission_start.cpp]: `reveal_destination( , false)` still isn't being used. The flag should be removed.

File: src/mission_start.cpp [hunk 8]
```diff
@@ -1627,3 +1627,82 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+    const npc *p = g->find_npc( miss->get_npc_id() );
+    if( p == NULL ) {
+        debugmsg( "mission_start::infect_npc() couldn't find an NPC!" );
+        return;
+    }
+
+    const tripoint destination = reveal_destination( omter_id, true );
+    if( destination != overmap::invalid_tripoint ) {
+        const oter_id oter = overmap_buffer.ter( destination );
+        add_msg( _( "%s have marked the only %s known to them on your map" ),
```
Comment (codemime) [inline src/mission_start.cpp]: Missing dot at the end of the sentence.

File: src/mission_start.cpp [hunk 9]
```diff
@@ -1627,3 +1627,82 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+    if( overmap_buffer.reveal_route( source_road, dest_road ) ) {
+        add_msg( _( "%s marks as well the road that leads to it..." ),
+                 p->name.c_str() );
+    }
+}
+
+void reveal_target( mission *miss, const char *omter_id )
+{
+    const npc *p = g->find_npc( miss->get_npc_id() );
+    if( p == NULL ) {
```
Comment (codemime) [inline src/mission_start.cpp]: We use `nullptr` instead of `NULL` (which is an `int` macro).

File: src/mission_start.cpp [hunk 10]
```diff
@@ -1627,3 +1627,82 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+        add_msg( _( "%s have marked the only %s known to them on your map" ),
+                 p->name.c_str(), oter->name.c_str() );
+        miss->set_target( destination );
+        if( one_in( 3 ) ) {
+            reveal_route( miss, destination );
+        }
+    }
+}
+
+void reveal_any_target( mission *miss, std::vector<std::string> omter_ids )
```
Comment (codemime) [inline src/mission_start.cpp]: Should be `const std::vector<std::string> &omter_ids`. Otherwise it's copied.

File: src/mission_start.cpp [hunk 11]
```diff
@@ -1627,3 +1627,82 @@ void mission_start::ranch_bartender_4(mission *miss)
...
+    const tripoint source_road = overmap_buffer.find_closest( source, "road", 3, false );
+    const tripoint dest_road = overmap_buffer.find_closest( destination, "road", 3, false );
+
+    if( overmap_buffer.reveal_route( source_road, dest_road ) ) {
+        add_msg( _( "%s marks as well the road that leads to it..." ),
+                 p->name.c_str() );
+    }
+}
+
+void reveal_target( mission *miss, const char *omter_id )
```
Comment (codemime) [inline src/mission_start.cpp]: I'd suggest `std::string` for the second param.

**Comment by codemime** [comment]
It's worth further JSONizing in order to:
- Prevent proliferation of hardcoded functions that all look alike.
- Enable other mission starts along with revealing targets.

Something like (either, but both should be supported):
```json
"revealed_target" : "station_radio",
"revealed_target" : ["office_doctor", "hospital"]
```

**Comment by patternoia** [comment]
Yes that is exactly my intent for further development

**Pull request merged**